src/cljdoc/util/sqlite_cache.clj

```diff
@@ -0,0 +1,188 @@
+
+  This namespace exposes `memo-sqlite` function which takes a
+  function to memoize and cache-spec, it retuns the memoized function.
+  This function uses the `SQLCache` for memoization.
+  "
+  (:require [clojure.core.cache :as cache]
+            [clojure.core.memoize :as memo]
+            [taoensso.nippy :as nippy]
+            [clojure.java.jdbc :as sql])
+  (:import (java.time Instant)))
+
+(def query-templates
+  {:fetch   "SELECT * FROM %s WHERE prefix = ? AND %s = ?"
+   :evict   "DELETE FROM %s WHERE prefix = ? AND %s = ?"
+   :cache   "INSERT OR IGNORE INTO %s (prefix, cached_ts, ttl ,%s, %S) VALUES (?, ?, ?, ?, ?)"
+   :refresh "UPDATE %s SET cached_ts = ?, ttl = ?, %s = ? WHERE prefix = ? and %s = ?"})
+
+(defn stale?
+  "Tests if `ttl` has expired for a cached item.
+
+  When `ttl` is `nil` item is not stale.
+  Otherwise `ttl` is compared with `cached_ts`."
+  [{:keys [ttl cached_ts]}]
+  (if (nil? ttl)
+    false
+    (> (- (System/currentTimeMillis)
+          (.toEpochMilli (Instant/parse cached_ts)))
+       ttl)))
+
+(defn fetch-item!
+  "Performs lookup by querying on the cache table.
+  Retuens deserialized cached item."
+  [k {:keys [db-spec key-prefix deserialize-fn table key-col value-col]}]
+  (let [query (format (:fetch query-templates) table key-col)
+        row-fn #(some-> % (get (keyword value-col)) deserialize-fn)]
+    (sql/query db-spec
+               [query key-prefix (pr-str k)]
+               {:row-fn row-fn :result-set-fn first})))
+
+(defn fetch!
+  [k {:keys [db-spec key-prefix table key-col]}]
+  (let [query (format (:fetch query-templates) table key-col)]
+    (sql/query db-spec [query key-prefix (pr-str k)] {:result-set-fn first})))
+
+(defn refresh!
+  [k v {:keys [db-spec key-prefix table key-col value-col serialize-fn ttl]}]
+  (let [query (format (:refresh query-templates) table value-col key-col)
+        value (serialize-fn @v)]
+    (sql/execute! db-spec [query (Instant/now) ttl value key-prefix (pr-str k)])))
+
+(defn cache!
+  [k v {:keys [db-spec key-prefix serialize-fn table key-col value-col ttl]}]
+  (let [query (format (:cache query-templates) table key-col value-col)
+        value (serialize-fn @v)]
+    (sql/execute! db-spec [query key-prefix (Instant/now) ttl (pr-str k) value])))
+
+(defn evict!
+  [k {:keys [db-spec key-prefix table key-col]}]
+  (let [query (format (:evict query-templates) table key-col)]
+    (sql/execute! db-spec [query key-prefix (pr-str k)])))
+
+(defn seed!
+  [{:keys [db-spec key-prefix table key-col value-col]}]
+  (let [column-specs [[:ttl "integer"]
+                      [:prefix "text"]
+                      [:cached_ts "text"]
+                      [(keyword key-col) "text"]
+                      [(keyword value-col) "text"]]
+        query (sql/create-table-ddl table
+                                    column-specs
+                                    {:conditional? true})]
+    (sql/execute! db-spec [query])))
```

Comment:
If the table is created like this there won't be any indexes except for row IDs — but our primary access is via the  `prefix` and `$key-col` columns. What I did in other tables and what is also appropriate here is to add a [`UNIQUE` constraint](https://www.techonthenet.com/sqlite/unique.php) to ensure there's no two rows with the same `prefix` and `$key-col`. This will implicitly add an index on the columns referenced as part of the `UNIQUE` constraint and therefore speed up queries significantly.

Also make sure that `prefix` and `$key-col` cannot be `null` since they are mandatory. Maybe even more columns should be enforced to have a value.

Reply:
Good point about the indexes. I was about to ask you about that. As you suggested I have added a `Unique constraint` on (`prefix`, `$key-col`) also added `not nulls` on `prefix` , `$key-col` and `cached_ts`.